Mathlib/Data/Finite/Defs.lean

```diff
@@ -0,0 +1,135 @@
+#align finite Finite
+
+theorem finite_iff_exists_equiv_fin {α : Sort _} : Finite α ↔ ∃ n, Nonempty (α ≃ Fin n) :=
+  ⟨fun ⟨e⟩ => ⟨_, ⟨e⟩⟩, fun ⟨_, ⟨e⟩⟩ => ⟨e⟩⟩
+#align finite_iff_exists_equiv_fin finite_iff_exists_equiv_fin
+
+theorem Finite.exists_equiv_fin (α : Sort _) [h : Finite α] : ∃ n : ℕ, Nonempty (α ≃ Fin n) :=
+  finite_iff_exists_equiv_fin.mp h
+#align finite.exists_equiv_fin Finite.exists_equiv_fin
+
+theorem Finite.of_equiv (α : Sort _) [h : Finite α] (f : α ≃ β) : Finite β := by
+  cases' h with n e
+  exact Finite.intro (f.symm.trans e)
+#align finite.of_equiv Finite.of_equiv
+
+theorem Equiv.finite_iff (f : α ≃ β) : Finite α ↔ Finite β :=
+  ⟨fun _ => Finite.of_equiv _ f, fun _ => Finite.of_equiv _ f.symm⟩
+#align equiv.finite_iff Equiv.finite_iff
+
+theorem Function.Bijective.finite_iff {f : α → β} (h : Bijective f) : Finite α ↔ Finite β :=
+  (Equiv.ofBijective f h).finite_iff
+#align function.bijective.finite_iff Function.Bijective.finite_iff
+
+theorem Finite.ofBijective [Finite α] {f : α → β} (h : Bijective f) : Finite β :=
+  h.finite_iff.mp ‹_›
+#align finite.of_bijective Finite.ofBijective
+
+instance [Finite α] : Finite (PLift α) :=
+  Finite.of_equiv α Equiv.plift.symm
+
+instance {α : Type v} [Finite α] : Finite (ULift.{u} α) :=
+  Finite.of_equiv α Equiv.ulift.symm
+
+/-- A type is said to be infinite if it is not finite. Note that `Infinite α` is equivalent to
+`isEmpty (Fintype α)` or `isEmpty (Finite α)`. -/
+class Infinite (α : Sort _) : Prop where
+  /-- assertion that `α` is `¬Finite`-/
+  notFinite : ¬Finite α
+#align infinite Infinite
+
+@[simp]
+theorem notFinite_iff_infinite : ¬Finite α ↔ Infinite α :=
+  ⟨Infinite.mk, fun h => h.1⟩
+#align not_finite_iff_infinite notFinite_iff_infinite
+
+@[simp]
+theorem not_infinite_iff_finite : ¬Infinite α ↔ Finite α :=
+  notFinite_iff_infinite.not_right.symm
+#align not_infinite_iff_finite not_infinite_iff_finite
+
+theorem Equiv.infinite_iff (e : α ≃ β) : Infinite α ↔ Infinite β :=
+  notFinite_iff_infinite.symm.trans <| e.finite_iff.not.trans notFinite_iff_infinite
+#align equiv.infinite_iff Equiv.infinite_iff
+
+instance [Infinite α] : Infinite (PLift α) :=
+  Equiv.plift.infinite_iff.2 ‹_›
+
+instance {α : Type v} [Infinite α] : Infinite (ULift.{u} α) :=
+  Equiv.ulift.infinite_iff.2 ‹_›
+
+theorem finite_orInfinite (α : Sort _) : Finite α ∨ Infinite α :=
+  or_iff_not_imp_left.2 <| notFinite_iff_infinite.1
+#align finite_or_infinite finite_orInfinite
+
+/-- `Infinite α` is not `Finite`-/
+theorem notFinite (α : Sort _) [Infinite α] [Finite α] : False :=
+  @Infinite.notFinite α ‹_› ‹_›
+#align not_finite notFinite
+
+protected theorem Finite.false [Infinite α] (_ : Finite α) : False :=
+  notFinite α
+#align finite.false Finite.false
+
+protected theorem Infinite.false [Finite α] (_ : Infinite α) : False :=
+  @Infinite.notFinite α ‹_› ‹_›
+#align infinite.false Infinite.false
+
+alias not_infinite_iff_finite ↔ Finite.of_not_infinite Finite.not_infinite
```

Comment:
Should we prove `finite.not_infinite` without `classical.choice` (both here and in Lean 3)?

Reply:
Maybe I don't understand where we have a `classical` dependency? I'm inclined to keep the port as is, but if want to change `mathlib` and port, no objections.

Reply:
`not_infinite_iff_finite` uses `not_not` which uses `classical.choice`. Don't worry about it - we can fix this later.